src/interceptors/Socket/SocketInterceptor.ts

```diff
@@ -0,0 +1,450 @@
+import net from 'node:net'
+import { randomUUID } from 'node:crypto'
+import { Readable } from 'node:stream'
+import { invariant } from 'outvariant'
+import { until } from '@open-draft/until'
+import { Interceptor } from '../../Interceptor'
+import {
+  InteractiveRequest,
+  toInteractiveRequest,
+} from '../../utils/toInteractiveRequest'
+import { emitAsync } from '../../utils/emitAsync'
+
+const { HTTPParser } = process.binding('http_parser')
+
+export interface SocketEventMap {
+  request: [
+    args: {
+      requestId: string
+      request: InteractiveRequest
+    },
+  ]
+  response: [
+    args: {
+      requestId: string
+      request: Request
+      response: Response
+      isMockedResponse: boolean
+    },
+  ]
+}
+
+export class SocketInterceptor extends Interceptor<SocketEventMap> {
+  static symbol = Symbol('socket')
+
+  constructor() {
+    super(SocketInterceptor.symbol)
+  }
+
+  protected setup(): void {
+    const self = this
+    const originalConnect = net.Socket.prototype.connect
+
+    net.Socket.prototype.connect = function (normalizedOptions) {
```

Comment:
Should this be `...normalizedOptions` here, if it's passed on as arguments to connect?

Reply:
@gioragutt, I think it should be but Node.js provides here a single normalized object of options. Let's dive into the source for `net.Socket.connect` to verify that's the case.

Reply:
If that's the case, how is the spread in line 44 working? JS syntax wise it wouldn't make sense otherwise

Reply:
I write about this here:

https://github.com/mswjs/interceptors/blob/a755028997ffd8f2a02e91b316e67ec9084bdb51/src/interceptors/Socket/SocketInterceptor.ts#L61-L73

TL;DR Node.js will normalize the args for `.connect()` _in some cases_ but not the others. That's why we get different type of args if the socket connection is called as a part of `http.request()` and raw like `new Socket().connect()`.